Restructure Protobuf-based message implementation

sergefdrv opened the pull request:
This pull request restructures Protobuf-based message implementation in order to solve its two main problems: non-determinism and possible misinterpretation of messages' authenticated payload. Check individual commit messages for more details.

Resolves #6

Commit:
messages/protobuf: Regenerate Protobuf code

Regenerate the outdated Protobuf code with protoc v3.11.2.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Move generated Protobuf code into sub-package

The package's documentation is cluttered with auto-generated stuff
which is not a part of packages' public API. Separate auto-generated
code into a sub-package.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

ynamiki commented on messages/protobuf/impl.go:
```diff
@@ -32,7 +31,7 @@ func NewImpl() messages.MessageImpl {
 	msg := &pb.Message{}
 	if err := proto.Unmarshal(data, msg); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal message wrapper: %s", err)
+		return nil, errors.Wrap(err, "failed to unmarshal message wrapper")
```
As you may know, Go 1.13 introduces [new build-in error handling features](https://blog.golang.org/go1.13-errors) and there is a compatibility package, [golang.org/x/xerrors](https://github.com/golang/go/wiki/ErrorValueFAQ) for earlier versions. How about using the new feature?

To use the feature, just replace `%s` to `%w`:
```
fmt.Errorf("failed to unmarshal message wrapper: %w", err)
```

sergefdrv replied:
Thank you for pointing this out; I actually overlooked this new feature. The new feature looks almost the same as what `pkg/errors` provide. One notable difference is that `pkg/errors` automatically annotates errors with stack trace, which might be useful when analyzing problems and debugging. What do you think?

sergefdrv replied:
I checked more carefully and found that `xerrors` package in fact annotates errors with stack trace, as well. Then I think we should rather use it in favor of `pkg/error` package. 

The only question if we should give up on supporting Go 1.11 or use the compatibility package? Given that this is a component to be used by other projects, I tend to think of preserving better compatibility, even though it is less convenient.

ynamiki replied:
`golang.org/x/xerrors` supports Go 1.11 or later:
https://github.com/golang/xerrors/blob/master/go.mod

So we don't need to update our compatibility policy.

sergefdrv replied:
Should we also include such a `go` directive in our `go.mod`?

ynamiki replied:
Yes, I will do. Thanks for creating the issue.

Commit:
messages/protobuf: Use xerrors package

According to issue #106, the code base should eventually switch to use
`golang.org/x/xerrors` package if favor of built-in `fmt` and `errors`
packages. Switch this package to use `xerrors` before doing further
changes.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Store Protobuf structures by reference

It is better to avoid copying and duplicating whole Protobuf
structures over. Redefine structures implementing message interfaces
to store pointers to Protobuf structures.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Do not embed Protobuf structures

There is no benefit in promoting methods from generated Protobuf
structures. As opposed to embedding Protobuf structures, declaring
explicitly named fields allows to avoid method/field name conflicts.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Always initialize messages

There is no use of uninitialized message structures and the code is
simpler when message structures are always initialized with dedicated
constructor functions.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Use Prepare in Commit's Protobuf definition

Commit message virtually includes the corresponding Prepare message.
As opposed to decomposing the embedded Prepare into individual bits,
represent it as a single field.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Add message type tag into authenticated data

Message type should be covered in authenticated data to guarantee that
it cannot be misinterpreted. This resolves #6.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Pass UI counter when making test messages

This will simplify further improvements to unit tests.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Compute authenticated payload deterministically

Protobuf does not guarantee deterministic serialization, thus it is
not suitable to compute messages' signed/certified payload. Introduce
helper functions to obtain authenticated content from Protobuf
messages, and use those to implement SignedPayload/CertifiedPayload
methods.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Add helper to marshal messages

Liberate individual message type implementation from marshaling the
generic message wrapper. This also allows to bring both marshaling and
unmarshaling message type selectors close to each other.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Test certified message payload more rigorously

Check that individual bits of the embedded message are covered in
certified message payload.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Flatten Protobuf message definitions

Now inner definitions of Protobuf messages are not used to compute
authenticated payload, the definitions can be simplified.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Rename fields in Protobuf definitions

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Optimize converting abstract messages to Protobuf

Normally, different implementations of the message interfaces will not
be used together. Assuming that, try to avoid recreating Protobuf
messages from abstract interfaces.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
messages/protobuf: Add documentation comments

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

sergefdrv commented:
Summary of [changes](https://github.com/hyperledger-labs/minbft/compare/97045dbc76355063c33020557fc1d7f73146a10c..e9754f08ba494605d3207caa414b0ebb12dee72a):

- Replaced `pkg/errors` with `xerrors`

ynamiki reviewed (approved):
Thanks for update. Looks good to me.

nhoriguchi reviewed (approved):
Sorry for late response, the suggested change looks nice.

The non-determinism of protobuf's serialization was new to me.  As described in [godoc](https://godoc.org/github.com/golang/protobuf/proto#Buffer.SetDeterministic) `deterministic` flag doesn't guarantee stable serialization over time, so it's recommended to "define their own canonicalization specification and implement their own serializer rather than relying on this API." That's what `authen.go` does.
